Allow to honor PYTHONHOME environment variable for +python/dyn and +python3/dyn

splhack opened the pull request:
According to Python 2.7 and 3.5 source code, Python doesn't honor `PYTHONHOME` environment variable after `Py_SetPythonHome` call.

```
static char *default_home = NULL;

void
Py_SetPythonHome(char *home)
{
    default_home = home;
}

char *
Py_GetPythonHome(void)
{
    char *home = default_home;
    if (home == NULL && !Py_IgnoreEnvironmentFlag)
        home = Py_GETENV("PYTHONHOME");
    return home;
}
```

So, check `PYTHONHOME` environment variable first. It is required for `pythondll` and `pythonthreedll` options.

ZyX-I commented on src/if_python.c:
```diff
@@ -928,7 +928,12 @@ Python_Init(void)
 
 #ifdef PYTHON_HOME
+# ifdef DYNAMIC_PYTHON
+	if (mch_getenv((char_u *)"PYTHONHOME") == NULL)
```
I think that you should only guard the condition with an additional `#ifdef` and not repeat `Py_SetPythonHome` call.

splhack committed:
Allow to honor PYTHONHOME environment variable for +python/dyn and +python3/dyn

According to Python 2.7 and 3.5 source code, Python doesn't honor `PYTHONHOME` environment variable after `Py_SetPythonHome` call.

	static char *default_home = NULL;

	void
	Py_SetPythonHome(char *home)
	{
			default_home = home;
	}

	char *
	Py_GetPythonHome(void)
	{
			char *home = default_home;
			if (home == NULL && !Py_IgnoreEnvironmentFlag)
					home = Py_GETENV("PYTHONHOME");
			return home;
	}

So, check `PYTHONHOME` environment variable first. It is required for `pythondll` and `pythonthreedll` options.

Example:

	let $PYTHONHOME="/usr/local/Frameworks/Python.framework/Versions/2.7"
	set pythondll="/usr/local/Frameworks/Python.framework/Versions/2.7/Python"

splhack commented:
Hi Bram,

Could you review this patch in order to use if_python and if_python3 with pythondll and pythonthreedll options properly? For example, using the following .vimrc,

```
let $PYTHONHOME="/usr/local/Frameworks/Python.framework/Versions/2.7"
set pythondll=/usr/local/Frameworks/Python.framework/Versions/2.7/Python
```

Python tries to load modules from PYTHON_HOME, what was defined at compile time, without this patch. The output of this py command

```
:py import os; print(os.__file__)
```

Without this patch: (wrong path)

```
/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.pyc
```

With this patch: (the right path)

```
/usr/local/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.pyc
```

brammool commented:
Kazuki Sakamoto wrote:

> Could you review this patch in order to use if_python and if_python3
> with pythondll and pythonthreedll options properly? For example, using
> the following .vimrc,
> 
> ```
> let $PYTHONHOME="/usr/local/Frameworks/Python.framework/Versions/2.7"
> set pythondll=/usr/local/Frameworks/Python.framework/Versions/2.7/Python
> ```
> 
> Python tries to load modules from PYTHON_HOME, what was defined at
> compile time, without this patch. The output of this py command
> 
> ```
> :py import os; print(os.__file__)
> ```
> 
> Without this patch: (wrong path)
> 
> ```
> /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.pyc
> ```
> 
> With this patch: (the right path)
> 
> ```
> /usr/local/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.pyc
> ```

It's in the todo list.

## 

Wizards had always known that the act of observation changed the thing that
was observed, and sometimes forgot that it also changed the observer too.
            Terry Pratchett  -  Interesting times

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///
